libraries/standard/mqtt/src/iot_mqtt_serialize.c

```diff
@@ -1572,127 +1625,101 @@ IotMqttError_t _IotMqtt_DeserializePublish( _mqttPacket_t * pPublish )
             &_logHideAll,
             "Retain bit is %d.", pOutput->retain );
 
-    /* Check for QoS 2. */
-    if( UINT8_CHECK_BIT( publishFlags, MQTT_PUBLISH_FLAG_QOS2 ) == true )
+    status = _getIncomingPublishQOS( publishFlags, pOutput );
+
+    if( status == IOT_MQTT_SUCCESS )
     {
-        /* PUBLISH packet is invalid if both QoS 1 and QoS 2 bits are set. */
-        if( UINT8_CHECK_BIT( publishFlags, MQTT_PUBLISH_FLAG_QOS1 ) == true )
+        IotLog( IOT_LOG_DEBUG,
+                &_logHideAll,
+                "QoS is %d.", pOutput->qos );
+
+        /* Parse the DUP bit. */
+        if( UINT8_CHECK_BIT( publishFlags, MQTT_PUBLISH_FLAG_DUP ) == true )
```

Comment:
Perhaps this and line 1622 could be in `_getIncomingPublishQOS`, so that all `UINT8_CHECK_BIT`s were in one function called `_checkPublishFlags`?

Reply:
OK that's good suggestion, that will reduce complexity as well.